hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenMintHandler.java

```diff
@@ -54,8 +77,174 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
+
+        // get the latest serial number minted for the token
+        var currentSerialNumber = token.lastUsedSerialNumber();
+        validateTrue((currentSerialNumber + metadataCount) <= MAX_NUM_ALLOWED, SERIAL_NUMBER_LIMIT_REACHED);
+
+        // Change the supply on token
+        changeSupply(token, treasuryRel, metadataCount, FAIL_INVALID, accountStore, tokenStore, tokenRelStore);
+
+        final var mintedSerials = new ArrayList<Long>(metadata.size());
+
+        // for each serial number minted increment serial numbers and create new unique token
+        for (final var meta : metadata) {
+            currentSerialNumber++;
+            // The default sentinel account is used (0.0.0) to represent unique tokens owned by the treasury
+            final var uniqueToken =
+                    buildNewlyMintedNft(treasuryAccount, consensusTime, tokenId, meta, currentSerialNumber);
+            nftStore.put(uniqueToken);
+            // all minted serials should be added to the receipt
+            mintedSerials.add(currentSerialNumber);
+        }
+        // Update last used serial number and number of owned nfts and put the updated token and treasury
+        // into the store
+        final var copyToken = token.copyBuilder();
+        final var copyTreasury = treasuryAccount.copyBuilder();
+
+        copyToken.lastUsedSerialNumber(currentSerialNumber);
+        copyTreasury.numberOwnedNfts(treasuryAccount.numberOwnedNfts() + metadataCount);
+        copyTreasury.headNftSerialNumber(currentSerialNumber);
+        copyTreasury.headNftId(tokenId.tokenNum());
+
+        tokenStore.put(copyToken.build());
+        accountStore.put(copyTreasury.build());
+
+        return mintedSerials;
+    }
+
+    /**
+     * Builds a new unique token when minting a non-fungible token.
+     * @param treasuryAccount - the treasury account
+     * @param consensusTime - the consensus time of the transaction
+     * @param tokenId - the token id
+     * @param meta - the metadata of the nft
+     * @param currentSerialNumber - the current serial number of the nft
+     * @return - the newly built nft
+     */
+    @NonNull
+    private Nft buildNewlyMintedNft(
+            @NonNull final Account treasuryAccount,
+            @NonNull final Instant consensusTime,
+            @NonNull final TokenID tokenId,
+            @NonNull final Bytes meta,
+            final long currentSerialNumber) {
+        return Nft.newBuilder()
+                .id(UniqueTokenId.newBuilder()
+                        .tokenTypeNumber(tokenId.tokenNum())
+                        .serialNumber(currentSerialNumber)
+                        .build())
+                .ownerNumber(0L)
+                .mintTime(Timestamp.newBuilder()
+                        .seconds(consensusTime.getEpochSecond())
+                        .nanos(consensusTime.getNano())
+                        .build())
+                .metadata(meta)
+                .ownerNextNftId(UniqueTokenId.newBuilder()
```

Comment:
Harmless to include these, but unnecessary per above.

Reply:
Thanks Michael. 🙌 Will fix this in my next PR